pystiche/papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,226 @@
+    "ulyanov_et_al_2016_dataset",
+    "ulyanov_et_al_2016_image_loader",
+    "ulyanov_et_al_2016_training",
+    "ulyanov_et_al_2016_images",
+    "ulyanov_et_al_2016_stylization",
+]
+
+
+def ulyanov_et_al_2016_transformer_optim_loop(
+    image_loader: DataLoader,
+    device: torch.device,
+    transformer: nn.Module,
+    criterion: nn.Module,
+    criterion_update_fn: Callable[[torch.Tensor, nn.ModuleDict], None],
+    get_optimizer: ulyanov_et_al_2016_optimizer,
+    impl_mode: str = "paper",
+    mode: str = "texture",
+    quiet: bool = False,
+    logger: Optional[OptimLogger] = None,
+    log_fn: Optional[
+        Callable[[int, Union[torch.Tensor, pystiche.LossDict], float, float], None]
+    ] = None,
+) -> nn.Module:
+
+    if logger is None:
+        logger = OptimLogger()
+
+    if log_fn is None:
+        log_fn = default_transformer_optim_log_fn(logger, len(image_loader))
+
+    optimizer = get_optimizer(transformer, impl_mode=impl_mode)
+
+    loading_time_start = time.time()
+    for batch, input_image in enumerate(image_loader, 1):
+        input_image = input_image.to(device)
+
+        if mode == "style":
+            criterion_update_fn(input_image, criterion)
+
+        loading_time = time.time() - loading_time_start
+
+        def closure():
+            processing_time_start = time.time()
+
+            optimizer.zero_grad()
+
+            output_image = transformer(input_image)
+            loss = criterion(output_image)
+            for key in loss.keys():
+                loss[key] = loss[key] / output_image.size()[0]
+            loss.backward()
+
+            processing_time = time.time() - processing_time_start
+
+            if not quiet:
+                batch_size = input_image.size()[0]
+                image_loading_velocity = batch_size / loading_time
+                image_processing_velocity = batch_size / processing_time
+                log_fn(batch, loss, image_loading_velocity, image_processing_velocity)
+
+            return loss
+
+        optimizer.step(closure)
+        loading_time_start = time.time()
+
+        # change learning rate during training
+
+        if impl_mode == "texturenetsv1":
+            if batch % 300 == 0:
+                for param_group in optimizer.param_groups:
+                    param_group["lr"] = param_group["lr"] * 0.8
+
+        elif impl_mode == "paper":
+            if batch % 200 == 0 and batch >= 1000:
+                for param_group in optimizer.param_groups:
+                    param_group["lr"] = param_group["lr"] * 0.7
+
+        elif impl_mode == "master":
+            if batch % 2000 == 0:
+                for param_group in optimizer.param_groups:
+                    param_group["lr"] = param_group["lr"] * 0.8
+        else:
+            raise NotImplementedError
+
+    return transformer
+
+
+def ulyanov_et_al_2016_training(
+    content_image_loader: DataLoader,
+    style: Union[str, torch.Tensor],
+    impl_mode: str = "paper",
```

Comment:
We should stick to `impl_params: bool` and `instance_norm: bool`. What exactly both parameters mean should be part of the documentation.